quinn-proto/src/connection/mod.rs

```diff
@@ -796,6 +797,44 @@ impl Connection {
                 break;
             }
 
+            // Send an off-path PATH_RESPONSE. LIFO since old challenges might have been forgotten by the sender.
+            if space_id == SpaceId::Data
+                && num_datagrams == 1
+                && self
+                    .path_responses
+                    .last()
+                    .map_or(false, |(_, remote)| *remote != self.path.remote)
+            {
+                let (response, remote) = self.path_responses.pop().unwrap();
+                let mut builder = builder_storage.take().unwrap();
+                trace!("PATH_RESPONSE {:08x} (off-path)", response.token);
+                buf.write(frame::Type::PATH_RESPONSE);
+                buf.write(response.token);
+                self.stats.frame_tx.path_response += 1;
+                builder.pad_to(MIN_INITIAL_SIZE);
+                let pn = builder.exact_number;
+                builder.finish_and_track(
+                    now,
+                    self,
+                    Some(SentFrames {
+                        non_retransmits: true,
+                        ..SentFrames::default()
+                    }),
+                    buf,
+                );
+                self.stats.udp_tx.datagrams += 1;
+                self.stats.udp_tx.ios += 1;
+                self.stats.udp_tx.bytes += buf.len() as u64;
+                self.path.congestion.on_sent(now, buf.len() as u64, pn);
+                return Some(Transmit {
+                    destination: remote,
+                    size: buf.len(),
+                    ecn: None,
+                    segment_size: None,
+                    src_ip: self.local_ip,
+                });
```

Comment:
This looks like it is duplicating a decent amount of stuff. Do you think it's worth deduplicating this a bit more?

Reply:
Factored out the stats stuff and did some tweaking. I feel like what's left is pretty distinctive to this case, though I'm open to ideas if you still think it could be more concise.

Reply:
This is probably okay for now -- maybe it some point it makes sense to wrap a builder around `Transmit`...